golang/Chapter06/model/client.go

```diff
@@ -0,0 +1,755 @@
+			_, err := pipeline.Exec()
+			if err != nil {
+				log.Println("pipeline err in AutoCompleteOnPrefix: ", err)
+				return err
+			}
+			res := tmp.Val()
+			if len(res) != 0 {
+				items = res
+			}
+			return nil
+		}, zsetName)
+		if err != nil {
+			continue
+		}
+		break
+	}
+	var result []string
+	for _, item := range items {
+		if !strings.Contains(item, "{") {
+			result = append(result, item)
+		}
+	}
+	return result
+}
+
+func (c *Client) JoinGuild(guild, user string) {
+	c.Conn.ZAdd("members:"+guild, &redis.Z{Member: user, Score: 0})
+}
+
+func (c *Client) LeaveGuild(guild, user string) {
+	c.Conn.ZRem("members:"+guild, user)
+}
+
+func (c *Client) AcquireLock(lockname string, acquireTimeout float64) string {
+	identifier := uuid.NewV4().String()
+
+	end := time.Now().UnixNano() + int64(acquireTimeout*1e6)
+	for time.Now().UnixNano() < end {
+		if c.Conn.SetNX("lock:"+lockname, identifier, 0).Val() {
+			return identifier
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+	return ""
+}
+
+func (c *Client) PurchaseItemWithLock(buyerId, itemId, sellerId string) bool {
+	buyer := fmt.Sprintf("users:%s", buyerId)
+	seller := fmt.Sprintf("users:%s", sellerId)
+	item := fmt.Sprintf("%s.%s", itemId, sellerId)
+	inventory := fmt.Sprintf("inventory:%s", buyerId)
+
+	locked := c.AcquireLock("market:", 10)
+	defer c.ReleaseLock("market:", locked)
+	if locked == "" {
+		return false
+	}
+
+	var (
+		price float64
+		funds float64
+	)
+
+	resZscore := &redis.FloatCmd{}
+	resHget := &redis.StringCmd{}
+
+	pipe := c.Conn.TxPipeline()
+	if err := c.Conn.Watch(func(tx *redis.Tx) error {
+		resZscore = pipe.ZScore("market:", item)
+		resHget = tx.HGet(buyer, "funds")
+		if _, err := pipe.Exec(); err != nil {
+			log.Println("pipeline err in watch func of PurchaseItemWithLock: ", err)
+			return err
+		}
+		price = resZscore.Val()
+		funds, _ = strconv.ParseFloat(resHget.Val(), 64)
+		return nil
+	}); err != nil {
+		log.Println("tx err in PurchaseItemWithLock: ", err)
+		return false
+	}
+
+
+	if price == 0 || price > funds {
+		return false
+	}
+
+	pipe.HIncrBy(seller, "funds", int64(price))
+	pipe.HIncrBy(buyer, "funds", int64(-price))
+	pipe.SAdd(inventory, itemId)
+	pipe.ZRem("market:", item)
+	if _, err := pipe.Exec(); err != nil {
+		log.Println("pipeline failed in PurchaseItemWithLock: ", err)
+		return false
+	}
+	return true
+}
+
+func (c *Client) ReleaseLock(lockname, identifier string) bool {
+	lockname = "lock:" + lockname
+	var flag = true
+	for flag {
+		err := c.Conn.Watch(func(tx *redis.Tx) error {
+			pipe := tx.TxPipeline()
+			if tx.Get(lockname).Val() == identifier {
+				pipe.Del(lockname)
+				if _, err := pipe.Exec(); err != nil {
+					return err
+				}
+				flag = true
+				return nil
+			}
+
+			tx.Unwatch()
+			flag = false
+			return nil
+		})
+
+		if err != nil {
+			log.Println("watch failed in ReleaseLock, err is: ", err)
+			return false
+		}
+
+		if !flag {
+			break
+		}
```

Comment:
This block seems extra, given the loop variable.

Reply:
Fix it, delete the extra if statement, and test ok! Thank you for the advice.